### PR TITLE
[WIP][BEAM-7336] Add Beam schema inferring for KafkaIO

### PR DESCRIPTION
**Pull request by iemejia**
As discussed offline, here it is a version of my proposed changes to #10966 
This one assumes that the user will extract the Keys and Values as it already the case in KafkaIO and allows to configure Beam Schema inference separately.
A nice improvement is the resolution using `SchemaRegistry` but I found some points there that probably should be addressed in a different issue.

R: @aromanenko-dev @reuvenlax

### CODE REVIEW

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -913,16 +985,55 @@ public void setValueDeserializer(String valueDeserializer) {
...
+              : getValueDeserializerProvider().getCoder(coderRegistry);
+      if (isValueInferBeamSchema()) {
+        return inferSchemaCoder(valueCoder, schemaRegistry);
+      }
+      return valueCoder;
+    }
+
+    @VisibleForTesting
+    static <T> SchemaCoder<T> inferSchemaCoder(Coder<T> coder, SchemaRegistry schemaRegistry) {
+      // At the moment there is not a general way to infer SchemaCoders from non-schema Coders (that
```
Comment (iemejia) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: @reuvenlax I tried to get the Avro resolution by registering `AvroRecordSchema` as a a new `SchemaRegistryProvider` but this produced a lot of issues. 

I have the impression we need to do something like what I mention to register which basic Coders we can support also as SchemaCoders e.g. Avro, Protobuf and others that contain a `Schema` and the `from/toRowFunction`s

**Comment by iemejia** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
Run Spotless PreCommit

**Comment by aromanenko-dev** [comment]
@reuvenlax Could you take a look on this?

**Comment by pabloem** [comment]
hi Alexey. I'm sorry this was left to languish... is this something you think can be used?

**Comment by aromanenko-dev** [comment]
@pabloem I'm pretty sure that this feature is still worth to have and these changes seems feasible at the first glance but I don't remember why it was not merged. 
@iemejia Do you recall why it was closed?

**Pull request closed without merging**